### DISCUSSION TITLE
Welcome to Box2D Discussion

### DISCUSSION THREAD

**Discussion by erincatto** [erincatto/box2d#693] [General]
Welcome to Box2D discussions.

Please post long form questions and feedback here for public discussion. If you have a bug report or a small feature request you may file and issue directly

**Please start a new topic as needed.**

**Comment by simonvanbernem** [comment]
Hi!

My name is Simon and I'm using box2d to make a 2D multiplayer space game with realtime physics, that is loosely inspired by FTL. You build ships, you shoot them into pieces, you scavange other ships for their parts and so on. The movement is entirely physics-based and the ships can be arbritrary polygonal shapes, which leads to some interesting problems.

This is also why I'm super thankful for the engine and the and the community! I didn't have a lot of knowledge about physics engines prior to working with box2d, and being able to just have a robust engine at hand and ask questions to people also interested in the topic is super helpful! You can get stuck quite hard when developing alone, so another set of eyes/ears is invaluable. Being able to look at the source also helps: It enabled me to extend box2d in places, spot performance problems and get a better understanding about physics engines in general.

I'm glad that you made this engine Erin, and that this community exists!

Two concrete things I'm doing with the engine right now that might be interesting:
- I added a hierarchical body-system on top of box2d. In my game, every spaceship is a single rigid body with many conceptual pieces (door, wall, engine, gun, floor, shield, ...), so thinking in terms of fixtures was not very intuitive. With the abstraction I can treat those things as individual bodies, attach/detach them from other bodies and set their absolute and relative position and rotation. The heart of the inner workings are the information about the hierarchie and a routine that transforms a set of fixtures. I was quite surprised how complicated the whole thing was to implement given that simple description.
- You can tell a spaceship to move in a certain direction in my game, and the spaceship will figure out how it has to actuate the (rocket-)engines it has, to make that movement happen (remember, all movement is physical, so the ship only accelerates through forces applied by the engines!). The engines might be at arbitrary positions and rotation on the ship, so this is non-trivial. I modelled this as a linear program and wrote a simplex-solver for it, which does pretty well. This becomes interesting in terms of physics, when AI comes into play: To make decisions, it has to predict where it will end up in the future. I did this prediction using newtons laws of motion, but discovered they were slightly off and would lead to the ship oscillating around the position it should come to rest at (because it overshot all the time). Turns out: This happenend due to box2d being a physics engine and not the real world: The time steps are discrete not continuous, and the prediction has to accomodate that. I was able to find a closed-form equation that can predict the position correctly at discrete time steps in the future for a given delta-time, which made the prediction error go down to insignificance. That was a really fun deep dive into the engine (I stepped through the integrator to understand what was happenening) and the math behind it, when constructing the closed form of the series!

Thats it for now. I'm looking forward to seeing some interesting stuff on here!

**Reply by thegrb93** [reply]
@simonvanbernem what's the name of the game? I've had a similar idea before and would be interested in checking it out.

**Reply by simonvanbernem** [reply]
The working title is "spaced". It's not public yet since I am still doing ground-work and have invested 0 hours into visuals so far. It's really more of a tech-demo currently, with partly unfinished tech.

**Comment by mgerhardy** [comment]
Hi, I'm using box2d for a classic 2D platformer (single- and multiplayer). 

http://www.caveproductions.org/

I've recently updated to box2d 2.4.1. If you are looking for things that must have been adopted, check out the commits at: https://github.com/mgerhardy/caveexpress.

I've updated the buoyancy controller by Chris Campbell to 2.4.1, too: https://github.com/mgerhardy/caveexpress/blob/master/src/caveexpress/server/entities/Buoyancy.h

**Reply by simonvanbernem** [reply]
FYI: I get to some sort of admin panel when following the first links. Looks like the site is a bit broken (I also get SSL warnings).

**Comment by PressStart10** [comment]
hey, can i  install via apt?

**Reply by PressStart10** [reply]
just wanna know

**Comment by mm-longcheng** [comment]
https://github.com/mm-longcheng/box2dc
Box2d is a great project. I port it by c. It's passed unit-test and testbed.  
I compare and checking code so many time, and need more unit tests.
The library size smaller, and runtime speed almost the same.
Box2d never use any stl, project clean and simple. But some time we need pure c for develop. So i do this.

**Reply by erincatto** [reply]
Very nice port! I'm currently working on v3 of Box2D here: https://github.com/erincatto/box2c

The goal is to have a pure C API. Most of the code is C but I'm doing a lot of multithreading, so I'm not sure if it will be all C internally.

**Comment by vkensou** [comment]
Maybe we can give b2AllocFcn and b2FreeFcn a attached parameter: userdata.

```
typedef void* b2AllocFcn( unsigned int size, int alignment, void* userdata );
typedef void b2FreeFcn( void* mem, void* userdata );
B2_API void b2SetAllocator( b2AllocFcn* allocFcn, b2FreeFcn* freeFcn, void* userdata );
```

I would prefer to transfer allocFcn and freeFcn to b2WorldDef. But it seems that too many changes are needed.

**Reply by erincatto** [reply]
Given that the allocators are global, then the userData could be global as well? So basically this shifts a global from user code to Box2D code.

I didn't want per world allocators because things like `b2DynamicTree` can be used without a world and need allocation. Are you wanting to use an arena allocation strategy?

**Reply by vkensou** [reply]
> Given that the allocators are global, then the userData could be global as well? So basically this shifts a global from user code to Box2D code.

If you don't want every world to have an allocator, then the userdata should also be global.

> I didn't want per world allocators because things like `b2DynamicTree` can be used without a world and need allocation. Are you wanting to use an arena allocation strategy?


Sometimes users may want to set up different allocators or pools for different systems, so that different strategies can be used. And if the user uses allocator like C++'s std:: allocator, passing in a userdata is required.

**Reply by erincatto** [reply]
I guess my point is that user data is only meaningful if it is per world/dynamic tree/etc. Otherwise you would have to have the same global in the calling code.

I'll think about the per world allocator. As I said, there are other things that can allocate. And this would require potentially a lot of plumbing.

Do you currently have a use case for having different allocators for different worlds? What problems would it solve for you?

**Reply by vkensou** [reply]
Honestly, I haven't quite figured out the specific use cases for allocation by worlds. I haven't used multiple worlds in box2d, so I can't definitively answer that question.

Just based on experience, whenever I create a big object, I tend to pass an allocator into it (using C++, I pass memory_desource). I just unlike the global state.

Maybe we can start by passing a userdata to the global allocator?

**Reply by vkensou** [reply]
make a pr
https://github.com/erincatto/box2d/pull/799

**Reply by erincatto** [reply]
I've mentioned a few times now that the user data as a global in Box2D doesn't make much sense. The pointer must already be a global in the client code. Context doesn't matter when it is global.

**Comment by spldm** [comment]
Hi. I found a bug, or maybe I don't understand something.

In world.c there is a code:
```
float fraction = worldContext->fcn( id, output.point, output.normal, output.fraction, worldContext->userContext );
worldContext->fraction = fraction;
```
Due to comments in samples user defined fcn can return -1 and it means that ray casting should ignore this shape. But we assign -1 to worldContext->fraction here. 
Then in the loop inside "b2World_CastRay" we have call stack:
b2World_CastRay->b2DynamicTree_RayCast -> RayCastCallback -> b2RayCastPolygon -> B2_ASSERT( b2IsValidRay( input ) ) where input.maxFraction equals -1 after user-defined fcn and so we have an error in input.

Is this a bug or am I missing something?

**Reply by erincatto** [reply]
You are correct. My test for this feature is against a purely static world. But a world with dynamic and static would fail on this. I'll fix this.

**Comment by spldm** [comment]
Hi.
I'm continuing to move from Box2D 2.x to 3.1 and I can't understand one problem already for a couple of days.
I got all contacts using b2World_GetContactEvents
Then loop over all "begin touch events":

```
b2ContactBeginTouchEvent& contact = contactEvents.beginEvents[i];
const b2Manifold& wld = contact.manifold;
const b2ManifoldPoint& point = wld.points[0];
```

The problem is that point.normalImpulse always equals 0. Always. Even if fast object hit the ground. Why could this be?
I do it immediately after b2World_Step.

Update: 
However, if I get contact data through b2Shape_GetContactData from shapes received from b2ContactBeginTouchEvent, this data contains correct values at the same time. So, it looks like the problem is only in data inside b2ContactBeginTouchEvent, which seems a problem, because you wrote in manual that it is not good in terms of performance to get this data from b2Shape_GetContactData.
The same situation is for maxNormalImpulse and normalVelocity.

**Reply by erincatto** [reply]
The begin touch event data is recorded before the constraint solver is engaged. So the impulses will always be zero. The workflow to get both the begin events and the final contact data is shown in the `ContactEvent` sample.

For context, it is not good to ask all shapes for contact data after all time steps. However, if you use events to narrow the scope then this is good for performance. This is shown in the `ContactEvent` sample.

**Reply by spldm** [reply]
Thank you for explanation. Maybe it'll be good to add it to the documentation, because it can confuse.

**Reply by spldm** [reply]
I learned ContactEvent example and it seems to me a bit suoptimal... 

If 2 objects make a collision we actually have only 1 collision. But we have to do b2Shape_GetContactData two times - for both objects.

Even worse: if 3 objects(A, B, C) collide at the same time, we have only 3 collision events. 
In stupid implementation we call b2Shape_GetContactData 6 times, each of them returns 2 collisions, so we have to process every collision 4 times: because, for example if we process collision A-B, calls b2Shape_GetContactData for A, we get A-B and A-C contact data. The same for B(A-B, B-C contact data) and repeat it two more times for A-C and B-C collision events.

In less stupid implementation we have to make a set, collect all objects in it and call b2Shape_GetContactData for them, but anyway in this way we get 2 copies of each collision from both sides.

What am I missing?

**Reply by erincatto** [reply]
The sample shows how to optimize with respect to using events rather than examining all shapes for new contacts every time step. Imagine a world with thousands of shapes, but you only get a few events per time step. Narrowing the scope using events should be much more efficient.

In the previous version of Box2D this could be accomplished with callbacks. But callbacks are bad for multithreading and they also are problematic because the user often wants to modify the physics world during the callback and this is not a safe operation. So version 3 switch to using event buffers where possible.

The `ContactEvent` sample shows how to get the contact data from both shapes, but you don't need to do this. You can get it from one shape because this information is shared. If you want to optimize at this level then the best practice is to ask both shapes for the contact capacity and get the contact data from the shape with smaller capacity.

I could add an function like `b2Shape_GetContactDataSpecific` that takes the `b2ShapeId` of another shape and only returns data for that (if it exists). This would be a bit more efficient.

**Reply by spldm** [reply]
> If you want to optimize at this level then the best practice is to ask both shapes for the contact capacity and get the contact data from the shape with smaller capacity.

Doesn't it possible a lot of simultaneous contacts? This is the main problem as I see. If 4 shapes(A, B, C, D) make contact events at the same time we have 6 contact events.

I can imagine a loop like this over b2ContactEvents::beginEvents:
First event is between shapes A and B, capacity of both is 3. we call b2Shape_GetContactData for A - 3 pieces of contact data.
Second event is between shapes B and C, capacity of both is 3, we call b2Shape_GetContactData for B - 3 pieces of contact data.
Third event is between shapes C and D, capacity of both is 3, we call b2Shape_GetContactData for C - 3 pieces of contact data.
etc...

In total we have 18 pieces of contact data. Of course I don't have to process them all by comparing shape ids, but I have to receive them all from Box2D. Isn't it too big overhead?

**Reply by erincatto** [reply]
No this is not a big deal and there is not a big cost there. Normally the thing to avoid is when one shape is a large ground object that may have hundreds of contacts. Dynamic bodies generally only have up to 3 or 4 contacts.

If you find this creates a performance problem for you then it may be worthwhile to implement something like `b2Shape_GetContactDataSpecific`. Let me know what you see in your benchmarks.

**Comment by spldm** [comment]
Hi, it's me again)
I have a lack of performance and profiler values are a bit strange, I didn't find the same in all your samples.
After b2World_GetProfile
I have next values:
step = 3.19
**pairs = 2.34**
collide = 0.19
solve = 0.65.
So, the main problem is "pairs", what could it mean? what should I pay attention to speed it up?
At that moment the counters(b2World_GetCounters) were: 1726 bodies, 4060 shapes, 183 contacts, 673 joints, 421 islands, 4 tasks  and tree height is 24

**Reply by erincatto** [reply]
The pairs section is finding potential overlaps. It probably means you are either creating lots of shapes and/or you have many moving bodies.

Also be careful not to spawn many overlapping shapes. This can happen if you create many bodies at the origin then teleport them after creation (b2Body_SetTransform).

**Reply by spldm** [reply]
I don't create many shapes during steps, but I really have a lot of moving bodies(most of them). 
And also some of them forms groups which are very close together. Could it be a problem too? How can I improve it? is there a way to mark some of them to say they can't overlap?

**Reply by erincatto** [reply]
You can use `b2Filter` within `b2ShapeDef` to make some shapes not collide. This is in the documentation here: https://box2d.org/documentation/group__shape.html#structb2_filter

Are you using multithreading? The pair tracking is able to use multiple threads.

**Reply by spldm** [reply]
No, I don't use multithreading, but I use custom callback function(b2World_SetCustomFilterCallback) to filter contacts.
